chunky/src/java/se/llbit/chunky/renderer/scene/Scene.java

```diff
@@ -893,6 +893,9 @@ public synchronized void loadChunks(TaskTracker taskTracker, World world, Collec
 
       ChunkPosition[] chunkPositions = chunksToLoad.toArray(new ChunkPosition[0]);
 
+      int[] cubeWorldBlocks = new int[16*16*16];
+      int[] cubeWaterBlocks = new int[16*16*16];
```

Comment:
We should change the Octree interface so that it can do such optimizations internally. The scene shouldn't need to care about this.

Reply:
It seemed better to me, the scene knows when it is setting blocks in bulk (loading) and it is setting blocks individually (finalization).
It is still possible to encapsulate this optimization in the octree (in fact that's what I did initially) but it adds quite a bit of complexity:
 - the octree has to keep a 16\*16\*16 buffer and remember the position this buffer has to be set
 -  every time `set` is called, we check that the set is in the same 16\*16\*16 cube than the previous, if not we actually do the bulk set, emptying the buffer. After that the block is added to the buffer
 - As we want `set` to set blocks individually during the finalization phase, we need a flag that says whether `set` operation should be accumulated for bulk set or done indiviidually, this flag is set to accumulate initially and changed when `startFinalization` is called.
 - If before finalization, some blocks are set in cube 1, then some blocks are set in cube 2 (so the buffer for cube 1 has been flushed and the blocks inserted) and then again some blocks in cube 1 are set, but not the same blocks that were set previously, when the insertion of those new blocks will be done, the insertion can either do as the `setCube` method and replace the whole content (which is bad because only some of the blocks where set so the other should have kept there value), or the buffer needs to also keep track of which block is set or not, and when the insertion is done, the blocks of the buffer that haven't been set need some handling to not lose the previous value (which is not good because this additional handling could be expensive).

So, as I see things, if this was implemented only inside the octree without change to the external interface, we have the choice between a erroneous way of doing things that happens to work today because the cubes are only being set once (but that's fragile) or a correct way that can't take full advantage of the optimization.

With a new method added to the interface like `setCube`, the semantic makes it obvious that the whole cube will be replaced by what we give it (because we can't give it a cube with some blocks are not set) so it can takes full advantages of the optimization without introducing errors when we need to set some blocks in cube while leaving the other untouched (the origianl `set` method is used for that).

Maybe there is another way of doing things that I haven't considered but this seems to be better to me, in fact I'd argue that the chunks should have control on how they insert themselves in the octree, this would make it possible for the future 32\*32\*32 chunks (from cubic chunks) to insert themselves in one operation, but that feels very non -OOP/separation of concerns.

Sorry for the big block of text, I feel like I had a hard time expressing myself and this may not be very clear, don't hesitate to ask any question.

Reply:
Don't worry, I really appreciate the detailed explanation. I was mostly worried about how this optimization (obviously) reduces the separation of concerns a bit. Then again I guess optimizations like this are not possible without doing so – the overhead you described probably makes it more complicated than one "special" set method.

Another way (that would keep these optimizations in the octree implementation): What would be cool (especially for the 32x32x32 chunks) would be if the _octree_ could decide how to insert the blocks. Maybe by passing a function that produces the chunk sections or individual blocks and then the octree builds itself from that? A bit like the BVH constructor that gets a list of entities (or primitives) and decides what to do with them by itself.